### PR TITLE
Audit usages of unicode case-changing methods.

### PR DESCRIPTION
**Pull request by frewsxcv**
Part of https://github.com/servo/servo/issues/17777.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17883)

### CODE REVIEW

**Comment by frewsxcv** [comment]
r? @SimonSapin 

N.B.: I didn't address all the instances of `to_uppercase`, `to_lowercase`, etc.

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -3875,17 +3875,16 @@ fn update_with_current_time_ms(marker: &Cell<u64>) {
...
-        "never" | "no-referrer" => Some(ReferrerPolicy::NoReferrer),
-        "default" | "no-referrer-when-downgrade" => Some(ReferrerPolicy::NoReferrerWhenDowngrade),
-        "origin" => Some(ReferrerPolicy::Origin),
-        "same-origin" => Some(ReferrerPolicy::SameOrigin),
-        "strict-origin" => Some(ReferrerPolicy::StrictOrigin),
-        "strict-origin-when-cross-origin" => Some(ReferrerPolicy::StrictOriginWhenCrossOrigin),
-        "origin-when-cross-origin" => Some(ReferrerPolicy::OriginWhenCrossOrigin),
-        "always" | "unsafe-url" => Some(ReferrerPolicy::UnsafeUrl),
-        "" => Some(ReferrerPolicy::NoReferrer),
+    match token {
```
Comment (frewsxcv) [inline components/script/dom/document.rs]: should https://docs.rs/cssparser/0.18.2/cssparser/macro.match_ignore_ascii_case.html be used here?

Comment (SimonSapin) [inline components/script/dom/document.rs]: We have a `match_ignore_ascii_case!` macro that does this more efficiently.

File: components/script/dom/htmlareaelement.rs [hunk 2]
```diff
@@ -240,13 +241,13 @@ impl HTMLAreaElement {
...
        let elem = self.upcast::<Element>();
        let shape = elem.get_string_attribute(&"shape".into());
-       let shp: Shape = match shape.to_lowercase().as_ref() {
-           "circle" => Shape::Circle,
-           "circ" => Shape::Circle,
-           "rectangle" => Shape::Rectangle,
-           "rect" => Shape::Rectangle,
-           "polygon" => Shape::Rectangle,
-           "poly" => Shape::Polygon,
+       let shp: Shape = match &shape {
```
Comment (SimonSapin) [inline components/script/dom/htmlareaelement.rs]: `match_ignore_ascii_case!` again

Comment (frewsxcv) [inline components/script/dom/htmlareaelement.rs]: Addressed in https://github.com/servo/servo/pull/17883/commits/719b09ca60920e91289b75550bfd0e572246d01d

File: components/script/dom/document.rs (hunk 1 above)
Comment (frewsxcv) [inline components/script/dom/document.rs]: Addressed in https://github.com/servo/servo/pull/17883/commits/719b09ca60920e91289b75550bfd0e572246d01d

**Comment by frewsxcv** [comment]
On a side note, that macro seems like a pretty helpful utility. Wonder if it's deserving of its own separate crate instead of (awkwardly) in the cssparser crate

**Comment by SimonSapin** [comment]
Thanks!

@bors-servo r+

----

Regarding the macro’s location, the thing is we have many such utilities and I don’t feel like dealing with the maintenance overhead of publishing many crates (especially if they’d be in many repositories).

**Comment by frewsxcv** [comment]
Which tests are failing here?

**Comment by SimonSapin** [comment]
There’s many, so I scripted a thing:
```sh
% curl -s http://build.servo.org/builders/linux-rel-wpt/builds/5243/steps/shell__2/logs/filtered-wpt-errorsummary.log/text | jq '{status: .status, test: .test}' -c|sort -u
```
```json
{"status":"CRASH","test":"/html/dom/elements/global-attributes/dataset-enumeration.html"}
{"status":"FAIL","test":"/custom-elements/reactions/DOMStringMap.html"}
{"status":"FAIL","test":"/html/dom/elements/global-attributes/custom-attrs.html"}
{"status":"FAIL","test":"/html/dom/elements/global-attributes/dataset-delete.html"}
{"status":"FAIL","test":"/html/dom/elements/global-attributes/dataset-get.html"}
{"status":"FAIL","test":"/html/dom/elements/global-attributes/dataset.html"}
{"status":"FAIL","test":"/html/dom/elements/global-attributes/dataset-set.html"}
{"status":"FAIL","test":"/html/dom/elements/global-attributes/data_unicode_attr.html"}
{"status":"FAIL","test":"/html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html"}
{"status":"FAIL","test":"/html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html"}
{"status":"FAIL","test":"/html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html"}
{"status":"FAIL","test":"/html/semantics/embedded-content/the-img-element/update-the-source-set.html"}
{"status":"TIMEOUT","test":"/dom/collections/domstringmap-supported-property-names.html"}
{"status":"TIMEOUT","test":"/html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html"}
```

**Comment by frewsxcv** [comment]
Latest force push has this change:

```diff
diff --git a/components/script/dom/htmlelement.rs b/components/script/dom/htmlelement.rs
index 05fcfcc8ba..1c1f8b8cbf 100644
--- a/components/script/dom/htmlelement.rs
+++ b/components/script/dom/htmlelement.rs
@@ -378,6 +378,7 @@ static DATA_HYPHEN_SEPARATOR: char = '\x2d';
 
 fn to_snake_case(name: DOMString) -> DOMString {
     let mut attr_name = String::with_capacity(name.len() + DATA_PREFIX.len());
+    attr_name.push_str(DATA_PREFIX);
     for ch in name.chars() {
         if ch.is_ascii_uppercase() {
             attr_name.push(DATA_HYPHEN_SEPARATOR);
```

@bors-servo try

**Comment by jdm** [comment]
```
▶ CRASH [expected OK] /html/dom/elements/global-attributes/dataset-enumeration.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ attempt to subtract with overflow (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }, at /home/servo/buildbot/slave/linux-rel-wpt/build/components/script/dom/htmlelement.rs:408)

  ▶ TIMEOUT [expected OK] /dom/collections/domstringmap-supported-property-names.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  └ attempt to subtract with overflow (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }, at /home/servo/buildbot/slave/linux-rel-wpt/build/components/script/dom/htmlelement.rs:408)
```

**Comment by frewsxcv** [comment]
Latest force push:

```diff
diff --git a/components/script/dom/htmlelement.rs b/components/script/dom/htmlelement.rs
index 1c1f8b8cbf..05683653fa 100644
--- a/components/script/dom/htmlelement.rs
+++ b/components/script/dom/htmlelement.rs
@@ -405,7 +405,7 @@ fn to_camel_case(name: &str) -> Option<DOMString> {
     if has_uppercase {
         return None;
     }
-    let mut result = String::with_capacity(name.len() - DATA_PREFIX.len());
+    let mut result = String::with_capacity(name.len().saturating_sub(DATA_PREFIX.len()));
     let mut name_chars = name.chars();
     while let Some(curr_char) = name_chars.next() {
         //check for hyphen followed by character
```

@bors-servo try

**Comment by SimonSapin** [comment]
@bors-servo r+

Thanks!

---

Reviewed 1 of 1 files at r1, 3 of 6 files at r2, 1 of 1 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17883#-:-KqWup2NnzWe8wW2ILez:b-pjnt64)*

**Pull request merged**